lib/tests/test_ecl_grid.cpp

```diff
@@ -0,0 +1,321 @@
+        z_depths.data(), NULL);
+    return grid;
+}
+
+TEST_CASE("Test dxv_dyv_dzv grids", "[unittest]") {
+    GIVEN("A grid constructed with dxv_dyv_dzv_depthz") {
+        int nx = 100;
+        int ny = 100;
+        int nz = 10;
+
+        double x_width = 1.0 / nx;
+        double y_width = 1.0 / ny;
+        double z_width = 3.0 / nz;
+
+        ecl_grid_type *grid = generate_dxv_dyv_dzv_depthz_grid(
+            100, 100, 10, x_width, y_width, z_width);
+
+        AND_GIVEN("Any cell in that grid") {
+            auto i = GENERATE(take(3, random(0, 99)));
+            auto j = GENERATE(take(3, random(0, 99)));
+            auto k = GENERATE(take(3, random(0, 9)));
+
+            THEN("get_xyz3 should return cell center in x, y plane") {
+                double xc, yc, zc;
+                ecl_grid_get_xyz3(grid, i, j, k, &xc, &yc, &zc);
+
+                REQUIRE_THAT(xc,
+                             WithinAbs(center_of_cell(x_width, i), 0.000001));
+                REQUIRE_THAT(yc, WithinAbs(center_of_cell(y_width, j), 1e-4));
+            }
+            THEN("The top cell corner has z offset by depth function") {
+                double xc, yc, zc;
+
+                ecl_grid_get_cell_corner_xyz3(grid, i, j, k, 0, &xc, &yc, &zc);
+
+                double x_center = center_of_cell(x_width, i);
+                double y_center = center_of_cell(y_width, j);
+                REQUIRE_THAT(
+                    zc,
+                    WithinAbs(z_width * k + depth_function(x_center, y_center),
+                              1e-4));
+            }
+
+            THEN("The bottom cell corner has z offset by depth function") {
+                double xc, yc, zc;
+
+                ecl_grid_get_cell_corner_xyz3(grid, i, j, k, 4, &xc, &yc, &zc);
+
+                double x_center = center_of_cell(x_width, i);
+                double y_center = center_of_cell(y_width, j);
+                REQUIRE_THAT(zc,
+                             WithinAbs(z_width * (k + 1) +
+                                           depth_function(x_center, y_center),
+                                       1e-4));
+            }
+
+            THEN("The volume should be simple multiplication") {
+                double volume = ecl_grid_get_cell_volume3(grid, i, j, k);
+                REQUIRE_THAT(volume,
+                             WithinAbs(z_width * x_width * y_width, 1e-4));
+            }
+        }
+
+        ecl_grid_free(grid);
+    }
+}
+
+/**
+ * Generates a grid, using the ecl_grid_alloc_GRDECL_kw constructor, with the
+ * x,y,z coordinates of the 0th corner of the ijkth cel is i,j,k. In other
+ * words this is a grid where the corners are on the whole numbers.
+ *
+ * Takes a vector of i,j,k,c,z tuples which changes the corner c of cell at
+ * i,j,k z value.
+ */
+ecl_grid_type *generate_coordkw_grid(
+    int num_x, int num_y, int num_z,
+    vector<tuple<int, int, int, int, double>> z_vector) {
```

Comment:
You're passing `z_vector` by copy, do it rather by `const &`.

Reply:
Fixed